flang/lib/Lower/OpenMP.cpp

```diff
@@ -50,22 +50,27 @@ static void genObjectList(const Fortran::parser::OmpObjectList &objectList,
 }
 
 template <typename Op>
-static void createBodyOfOp(Op &op, Fortran::lower::FirOpBuilder &firOpBuilder,
-                           mlir::Location &loc) {
-  if constexpr (std::is_same_v<Op, omp::WsLoopOp>) {
-    unsigned int numIVs = op.step().size();
-    SmallVector<Type, 8> argTypes(numIVs, firOpBuilder.getIndexType());
-    firOpBuilder.createBlock(&op.getRegion(), {}, argTypes);
+static void createBodyOfOp(Op &op, Fortran::lower::AbstractConverter &converter,
+                           mlir::Location &loc,
+                           const Fortran::semantics::Symbol *arg = nullptr) {
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  // If an argument for the region is provided then create the block with that
+  // argument. Also update the symbol's address with the mlir argument value.
+  // e.g. For loops the argument is the induction variable. And all further
+  // uses of the induction variable should use this mlir value.
+  if (arg) {
+    firOpBuilder.createBlock(&op.getRegion(), {}, {converter.genType(*arg)});
+    converter.setSymbolAddress(*arg, op.getRegion().front().getArgument(0));
```

Comment:
I had a look with unparse-with-symbols and I see that a new symbol is created for the loop variable. It is created in the following location.
https://github.com/llvm/llvm-project/blob/a9e68db9736080373d73606d89a270e38d7f1273/flang/lib/Semantics/resolve-directives.cpp#L1200

But it is not working without forced. Not sure what the issue is. Will have a look tomorrow.

Reply:
The semantic checks and symbol creation for OpenMP were not enabled in bbc. I fixed that with https://github.com/flang-compiler/f18-llvm-project/pull/652.

But i still need to use forced, even after this fix, and pushing the scope.

Reply:
When I try to add the symbol "i (OmpPrivate, OmpPreDetermined): HostAssoc" there is already a value there "%0 = fir.alloca i32 {name = "_QEi"}" for this symbol. This alloca value is for the variable which is outside the openmp region.

Does that mean the forced is required always. Note that by forced replacement here the replacement only happens inside the OpenMP region since the symbol is the privatised one.

Reply:
OK, something is weird. You really should not have to force the symbol here, especially when pushing a new scope. Can you update the PR with what you have so that I can have a look at what is happening in the symbolMap ?

Reply:
I have updated the PR.

Reply:
You definitely should be able to add a new binding in a new subscope for the symbol without having to force. Sounds like a bug.

Reply:
Symbols created due to OpenMP clauses like private or due to the semantics (like index of a worksharing loop) have HostAssociation set. In Fortran::lower::SymMap::lookupSymbol if there is a host-association then the host-associated symbol's value is returned. There is a "FIXME BUG" for this code.

What will be your recommendation here? Fix the lookup or fix the host-assoc setting for the OpenMP symbols?

https://github.com/flang-compiler/f18-llvm-project/blob/a976029095c382617b6a987ed4f6b29fbc81a745/flang/lib/Lower/SymbolMap.cpp#L200

Reply:
I think you can keep the `forced` that you have for now and had a `FIXME: removed forced when symbol lookup stop following host association`. I indeed need to change the way statement functions are handled to fix the FIXME you are pointing to and that is causing you to have to force the symbol in.

Reply:
Have added the fixme.

Reply:
No, it is not on my top priority list. Is this blocking you for something ? The "force" is not good practice/future-proof, but doesn't it allow you to get the the behavior you need ?

Reply:
We have another situation where indices of sequential loops inside a parallel region must be private. New symbols are created during semantic phase but because of host association the new symbol is not used. So kernels like matrix multiplication where the outermost loop is a parallel worksharing loop but the inner loops are sequential fails. 

Also, going forward we are going to implement the privatisation clauses and for this we need the new symbols to be picked.

If it is not in your priority list then I can if you can provide some more explanation and places where the current behaviour is required. Is it statement functions only?

Reply:
> If it is not in your priority list then I can if you can provide some more explanation

That would be great thanks !

> Is it statement functions only ?
Right now yes, blocks will require the same functionality, but they are not yet lowered.

Here is an illustration of the statement function case:

```
real function foo(x)
    real :: x, y, stmt_func
    stmt_func(y) = x + y
    foo = stmt_func(42.)
end function
```

If you use `f18 -fdebug-dump-symbols` on this, you will see that there is a subprogram scope for the statement function that looks like:

```
    Subprogram scope: stmt_func size=8 alignment=4
      stmt_func size=4 offset=0: ObjectEntity funcResult type: REAL(4)
      x: HostAssoc
      y size=4 offset=4: ObjectEntity dummy type: REAL(4)
```

The point of following host association in lowering was to allow `x` reference in the statement function expression to be lowered to its host associated symbol (the dummy argument `x` in my example).

That was easy and ok at the time, but this behavior does not match what is needed for OpenMP host associated symbols, so instead, I think the following code that lowers statement function should explicitly map statement function host associated symbols to their parent scope lowered symbol box:
https://github.com/flang-compiler/f18-llvm-project/blob/49f75bf752e3b354a43472f7de53ebcb858e468f/flang/lib/Lower/ConvertExpr.cpp#L1487
The statement function lowering is already pushing/ poping a symbol map scope, so there is nothing new to do to get rid of these symbol mapping after the statement function has been lowered.


I think mapping the symbols with HostAssociated attribute from `Fortran::evaluate::CollectSymbols(*expr)` to SymbolBox from their parent symbols should be enough (with some care to avoid trying to map a symbol twice if it appears multiple times in the expression). That is a quick idea that I did not try, their might be a few pitfalls. If the analysis 

After that, I think you could get rid of the part that is following host association in symbol map lookups and the "force" in "bindSymbol".